### PR TITLE
Downloader interface

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: src/downloader/downloader.hpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
+/*
```
Comment (al1img) [inline src/downloader/downloader.hpp:1]: add tag [donwloader] to commit message.

File: src/downloader/downloader.hpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+ */
+
+#ifndef DOWNLOADER_HPP_
+#define DOWNLOADER_HPP_
+
+#include "aos/common/downloader.hpp"
+#include "aos/common/error.hpp"
+#include "aos/common/string.hpp"
+
+namespace aos {
```
Comment (al1img) [inline src/downloader/downloader.hpp:15]: We do not use namespaces in zephyr app.

File: src/downloader/downloader.hpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+namespace downloader {
+
+/**
+ * @brief Image content request
```
Comment (al1img) [inline src/downloader/downloader.hpp:19]: Remove @brief tag from all comments.

File: src/downloader/downloader.hpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+ *
```
Comment (al1img) [inline src/downloader/downloader.hpp:20]: No need for this extra line in all comments.

File: src/downloader/downloader.hpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+ */
+struct ImageContentRequest {
```
Comment (al1img) [inline src/downloader/downloader.hpp:22]: Move to under interface it is used.

File: src/downloader/downloader.hpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    String   url;
+    uint64_t requestId;
```
Comment (al1img) [inline src/downloader/downloader.hpp:24]: `requestID`. Let's keep go way for acronyms.

File: src/downloader/downloader.hpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+ */
+class Downloader : public aos::downloader::DownloaderItf, public aos::downloader::DownloadReceiverItf {
+public:
+    /**
+     * @brief Initializes downloader instance.
+     *
+     * @param downloadRequester
+     * @return Error
+     */
+    aos::Error Init(aos::downloader::DownloadRequesterItf& downloadRequester);
```
Comment (al1img) [inline src/downloader/downloader.hpp:126]: Should be pointer to requester. You can't store refs.

File: src/downloader/downloader.hpp [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+     * @return Error
+     */
+    aos::Error ReceiveImageContentInfo(aos::downloader::ImageContentInfo& content) override;
+
+private:
+    aos::downloader::DownloadRequesterItf* mDownloadRequester {};
+    aos::downloader::ImageContentInfo      mImageContentInfo {};
+};
+
+#endif
```
Comment (al1img) [inline src/downloader/downloader.hpp:159]: New line at the end.
HINT for vs cde:
```
"files.insertFinalNewline": true,
"files.trimFinalNewlines": true,
```

File: src/downloader/downloader.hpp [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    String   contentType;
+};
+
+/**
+ * @brief Image content info
+ *
+ */
+struct Image {
+    String           relativePath;
+    StaticBuffer<32> sha256;
```
Comment (al1img) [inline src/downloader/downloader.hpp:34]: There is dedicated const in types.hpp for sha size.

File: src/downloader/downloader.hpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+namespace aos {
+namespace downloader {
+
+/**
+ * @brief Image content request
+ *
+ */
+struct ImageContentRequest {
+    String   url;
```
Comment (al1img) [inline src/downloader/downloader.hpp:23]: String doesn't hold the buffer. Should be StaticString<> here.

File: src/downloader/downloader.hpp [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+};
+
+/**
+ * @brief Image content info
+ *
+ */
+struct Image {
+    String           relativePath;
+    StaticBuffer<32> sha256;
+    uint64_t         size;
```
Comment (al1img) [inline src/downloader/downloader.hpp:35]: size_t for size

File: src/downloader/downloader.hpp [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+class DownloadRequesterItf {
+public:
+    /**
+     * @brief Send image content request
+     *
+     * @param request
+     * @param info
+     * @return Error
+     */
+    virtual Error SendIamgeContentRequest(const ImageContentRequest& request, ImageContentInfo& info) = 0;
```
Comment (al1img) [inline src/downloader/downloader.hpp:73]: I don't clearly get why we need two different struct in request?

File: src/downloader/downloader.hpp [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+     * Destroy the Download Requester Itf object
+     */
+    virtual ~DownloadRequesterItf() = default;
+};
+
+/**
+ * File info
+ */
+struct FileInfo {
+    aos::String                         relativePath;
```
Comment (Eabramov84) [inline src/downloader/downloader.hpp:49]: use stsitc string

File: src/downloader/downloader.hpp [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    uint64_t                            size;
+};
+
+/**
+ * Image content info
+ */
+struct ImageContentInfo {
+    uint64_t                       requestID;
+    aos::StaticArray<FileInfo, 32> files;
+    aos::String                    error;
```
Comment (Eabramov84) [inline src/downloader/downloader.hpp:60]: use static string

File: src/downloader/downloader.hpp [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+public:
+    Downloader() = default;
+
+    /**
+     * Initializes downloader instance.
+     *
+     * @param downloadRequester download requester instance.
+     * @return Error
+     */
+    aos::Error Init(DownloadRequesterItf* downloadRequester);
```
Comment (Eabramov84) [inline src/downloader/downloader.hpp:121]: use reference instead if pointer

File: src/downloader/downloader.hpp [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+ */
+
+#ifndef DOWNLOADER_HPP_
+#define DOWNLOADER_HPP_
+
+#include "aos/common/downloader.hpp"
+#include "aos/common/error.hpp"
+#include "aos/common/string.hpp"
+#include "aos/common/types.hpp"
+
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/downloader/downloader.hpp:15]: Done

File: src/downloader/downloader.hpp [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+};
+
+/**
+ * @brief Download receiver interface
+ *
+ */
+class DownloadReceiverItf {
+public:
+    /**
+     * Receive image content
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/downloader/downloader.hpp:85]: Done

File: src/downloader/downloader.hpp (hunk 15 above)
Comment (MykolaSolyanko) [reply src/downloader/downloader.hpp]: Done

File: src/downloader/downloader.hpp [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+ * Max file chunk size
+ */
+constexpr auto cMaxFileChunkSize = 1024;
+
+/**
+ * File chunk
+ */
+struct FileChunk {
+    uint64_t            requestID;
+    aos::String         relativePath;
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/downloader/downloader.hpp:72]: Done

**Review by Eabramov84 — APPROVED** [review]

File: src/downloader/downloader.hpp [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+#include "aos/common/downloader.hpp"
+#include "aos/common/error.hpp"
+#include "aos/common/string.hpp"
+#include "aos/common/types.hpp"
+
+/**
+ * Image content request.
+ */
+struct ImageContentRequest {
+    aos::String          url;
```
Comment (al1img) [inline src/downloader/downloader.hpp:20]: StaticString

Comment (MykolaSolyanko) [reply src/downloader/downloader.hpp]: Done

**Pull request merged**